File: setCounting.py
def countSets(array):
    numSets = 0
    for x in range(0, len(array) - 2):
        for y in range(x + 1, len(array) - 1):
            for z in range(y+1, len(array)):
                if(array[x]^array[y] == array[z]):
                    numSets+=1
    return(numSets)


#get dimension
try:
    dim = int(input("What dimension do you want? "))
    print("The dimension is:", dim)
except ValueError:
    print("Invalid input. Please enter an integer.")

numCards = 2**dim - 1 #total possible cards

#get number of cards
pickNum = numCards + 1
while(pickNum > numCards):
    try:
        pickNum = int(input("How many cards? "))
        if(pickNum <= numCards):
            print("Collections of size: ", pickNum)
        else:
            print("Too many cards! Max is:", numCards)
    except ValueError:
        print("Invalid input. Please enter an integer.")


    
numCollections = 0 #counts possible collections
cards = [] #array of (pickNum) cards
maxs = [] #last possible collection of cards
setCounts = [] #number of sets in each possible collection

#generate the starting cards and maxs array
#cards look like [1, 2, 3...]
#maxs look like [8, 9, 10...] (depends on size)
for num in range(1, pickNum+1):
    cards.append(num)
    maxs.append(numCards - (pickNum - num))
print("\n")
print("Starting at:", cards)
print("Ending at:", maxs)

#cycle through all possible collections of (pickNum) cards, counting the number of sets in each
while(cards[0] <= maxs[0]):
    setCounts.append(countSets(cards))
    numCollections +=1 
    
    #if last card isn't maxed out, increase it
    #otherwise, find the next collection
    if(cards[pickNum-1] < maxs[pickNum-1]):
        cards[pickNum-1] += 1
    elif(cards[pickNum-1 == maxs[pickNum-1]]):
        #find the last card that isn't maxed out
        index = pickNum - 1
        while(cards[index] == maxs[index]):
            index-=1
            if(index == -1):
                break
            
        if(index == -1):
            index+=1 #just to not go out of bounds
        cards[index] +=1 #increase the last card not maxed out
        while(index < pickNum-1):
            cards[index+1] = cards[index] + 1
            index +=1
            #"resets" all the cards after the index
            #eg, [1, 2, 14, 15] -> [1, 3, 4, 5]
    else:
        print("last card went out of range")
        

maxSetCount = max(setCounts) #highest number of sets in any possible collection
print("\n")
print("Max Set Count:", maxSetCount)
print("Number of possible collections:", numCollections)

setData = []
for x in range(0, maxSetCount+1):
    setData.append(0)

for x in setCounts:
    setData[x] += 1

print("\n")
print("--Set Data--")
print("#Sets  Frequency")
for x in range(0, maxSetCount+1):
    print(x, "\t"," ", setData[x])
 

    

File: setCounting.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
#include <algorithm>

using namespace std;

int countSets(const vector<int>& array) {
    int numSets = 0;
    for (size_t x = 0; x < array.size() - 2; ++x) {
        for (size_t y = x + 1; y < array.size() - 1; ++y) {
            for (size_t z = y + 1; z < array.size(); ++z) {
                if ((array[x] ^ array[y]) == array[z]) {
                    numSets++;
                }
            }
        }
    }
    return numSets;
}

int main() {
    int dim;
    cout << "What dimension do you want? ";
    cin >> dim;

    if (cin.fail()) {
        cout << "Invalid input. Please enter an integer." << endl;
        return 1;
    }
    cout << "The dimension is: " << dim << endl;

    int numCards = (1 << dim) - 1; // total possible cards

    int pickNum = numCards + 1;
    while (pickNum > numCards) {
        cout << "How many cards? ";
        cin >> pickNum;

        if (cin.fail()) {
            cout << "Invalid input. Please enter an integer." << endl;
            cin.clear();
            continue;
        }

        if (pickNum <= numCards) {
            cout << "Collections of size: " << pickNum << endl;
        } else {
            cout << "Too many cards! Max is: " << numCards << endl;
        }
    }

    int numCollections = 0; // counts possible collections
    vector<int> cards; // array of (pickNum) cards
    vector<int> maxs; // last possible collection of cards
    vector<int> setCounts; // number of sets in each possible collection

    // generate the starting cards and maxs array
    for (int num = 1; num <= pickNum; ++num) {
        cards.push_back(num);
        maxs.push_back(numCards - (pickNum - num));
    }
    cout << "\nStarting at: ";
    for (const auto& card : cards) cout << card << " ";
    cout << "\nEnding at: ";
    for (const auto& max : maxs) cout << max << " ";
    cout << endl;

    // cycle through all possible collections of (pickNum) cards, counting the number of sets in each
    while (cards[0] <= maxs[0]) {
        setCounts.push_back(countSets(cards));
        numCollections++;

        // if last card isn't maxed out, increase it
        // otherwise, find the next collection
        if (cards[pickNum - 1] < maxs[pickNum - 1]) {
            cards[pickNum - 1]++;
        } else if (cards[pickNum - 1] == maxs[pickNum - 1]) {
            // find the last card that isn't maxed out
            int index = pickNum - 1;
            while (cards[index] == maxs[index]) {
                index--;
                if (index == -1) {
                    break;
                }
            }

            if (index == -1) {
                index++; // just to not go out of bounds
            }
            cards[index]++; // increase the last card not maxed out
            while (index < pickNum - 1) {
                cards[index + 1] = cards[index] + 1;
                index++;
            }
        } else {
            cout << "last card went out of range" << endl;
        }
    }

    int maxSetCount = *max_element(setCounts.begin(), setCounts.end()); // highest number of sets in any possible collection
    cout << "\nMax Set Count: " << maxSetCount << endl;
    cout << "Number of possible collections: " << numCollections << endl;

    vector<int> setData(maxSetCount + 1, 0);
    for (const auto& count : setCounts) {
        setData[count]++;
    }

    cout << "\n--Set Data--" << endl;
    cout << "#Sets  Frequency" << endl;
    for (int x = 0; x <= maxSetCount; ++x) {
        cout << x << "\t " << setData[x] << endl;
    }

    return 0;
}
